[MM-13849] Display confirmation box for [at]all and [at]channel

hahmadia opened the pull request:
**Problem to be addressed**
I have attached a video showing the functionality of this feature. I just have a problem where the user counts doesn't update properly when a new user joins. Everything else updates accordingly except the user count in the room. Any advice on how to solve would be greatly appreciated.

### Summary
Adds a confirmation box when posting to `@all` or `@channel` on the mobile.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10185

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [ ] All new/modified APIs include changes to [mattermost-redux](https://github.com/mattermost/mattermost-redux) (please link)
- [ ] Has UI changes
- [X] Includes text changes and localization file updates

#### Device Information
This PR was tested on: OnePlus 5T, Android version 9, OxygenOS version: 9.0.7
*Aside* I unfortunately don't have access to Iphone to test this on :( How can I go about making sure it also works on Iphones? 🤔 

#### Screenshots
Demonstration video: https://streamable.com/af3l5
You can see that at the end of the video the number of people displayed should be 7 however it remains 6.

saturninoabril commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -408,62 +410,92 @@ export default class PostTextBoxBase extends PureComponent {
         } else {
+            const currentMembersCount = this.props.currentChannelMembersCount;
+            const notificationsToChannel = this.props.enableConfirmNotificationsToChannel;
+            const toAllorChannel = (/\B@(all|channel)\b/i).test(value);
```
I suggests to pull out this regex test whether the text contains valid [at]all and [@]channel and have it tested with several test cases similar to how we're doing in the webapp - https://github.com/mattermost/mattermost-webapp/blob/master/utils/post_utils.test.jsx#L14

hahmadia replied:
Resolved!

hahmadia commented:
Thank you all for your feedback. Much appreciated :)

@saturninoabril I have addressed the problems you mentioned.
1. Timezone text will now appear as long as the timezone flag is set through the admin panel.
2. Posting @all or @channel in a code or code block will not warn the user anymore
3. I have extracted the regex into it's own function and wrote test cases against it.

@esethna 
I was able to figure out the user channel not updating. It was due to an action (getChannelStats) that wasn't being called when the channel was mounted as well as when new users join the channel. It should now be working fine.

saturninoabril commented on app/components/combined_system_message/combined_system_message.js:
```diff
@@ -211,6 +213,8 @@ export default class CombinedSystemMessage extends React.PureComponent {
             this.props.actions.getMissingProfilesByUsernames(allUsernames);
         }
+
+        this.props.actions.getChannelStats(this.props.currentChannelId);
```
Are these necessary?  Same with the one at `channel_base.js`

hahmadia commented:
> @hahmadia Looking good.
> 
> Last item is the bold text on the "# people" & "# timezones" as mentioned by Eric above.
> ![Screen Shot 2019-07-31 at 4 12 34 PM](https://user-images.githubusercontent.com/5334504/62195350-66175f80-b3ae-11e9-9c90-2de98a8a86b6.png)

@saturninoabril I looked into that and explained @esethna that I can't bold the text because I am using Alert.alert which doesn't allow text bolding. Throughout the mobile app, we are using Alert.alert for confirmation and I was trying to keep it consistent. If we do require text bolding, it means that a custom transparent modal needs to be created which would make it inconsistent with the rest of the App. Please let me know your thoughts. Thanks

saturninoabril commented:
That makes sense.  If it's good with Eric, then all are now set.

saturninoabril commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -408,62 +414,154 @@ export default class PostTextBoxBase extends PureComponent {
+            notifyAllMessage = (
+                intl.formatMessage(
+                    {
+                        id: 'mobile.post_textbox.entire_channel.message',
```
This should have its own translation ID.

hahmadia replied:
Resolved! Thanks for pointing that out. Didn't catch it

enahum commented on app/components/combined_system_message/combined_system_message.js:
```diff
@@ -211,6 +213,8 @@ export default class CombinedSystemMessage extends React.PureComponent {
             this.props.actions.getMissingProfilesByUsernames(allUsernames);
         }
+
+        this.props.actions.getChannelStats(this.props.currentChannelId);
```
why fetching the stats on each `combined_system_message` that gets rendered? this does not look right to me

hahmadia replied:
@enahum After your suggestions given to me, I was able to pinpoint and found out that this is currently a problem in the web-app ( https://github.com/mattermost/mattermost-server/issues/11823 ) . I have removed this line of code. If users are added manually to a channel or a user joins a channel themself, then the `user_added` websocket action is broadcasted. Thanks for your help!

enahum commented on app/components/post_textbox/index.js:
```diff
@@ -42,8 +43,13 @@ function mapStateToProps(state, ownProps) {
     const currentUserId = getCurrentUserId(state);
     const status = getStatusForUserId(state, currentUserId);
     const userIsOutOfOffice = status === General.OUT_OF_OFFICE;
+    const enableConfirmNotificationsToChannel = config.EnableConfirmNotificationsToChannel === 'true';
```
```suggestion
    const enableConfirmNotificationsToChannel = config?.EnableConfirmNotificationsToChannel === 'true';
```

enahum commented on app/components/post_textbox/index.js:
```diff
@@ -42,8 +43,13 @@ function mapStateToProps(state, ownProps) {
     const userIsOutOfOffice = status === General.OUT_OF_OFFICE;
+    const enableConfirmNotificationsToChannel = config.EnableConfirmNotificationsToChannel === 'true';
+    const currentChannelStats = getCurrentChannelStats(state);
+    const currentChannelMembersCount = currentChannelStats && currentChannelStats.member_count;
```
```suggestion
    const currentChannelMembersCount = currentChannelStats?.member_count || 0;
```

enahum commented on app/components/post_textbox/index.js:
```diff
@@ -42,8 +43,13 @@ function mapStateToProps(state, ownProps) {
+    const enableConfirmNotificationsToChannel = config.EnableConfirmNotificationsToChannel === 'true';
+    const currentChannelStats = getCurrentChannelStats(state);
+    const currentChannelMembersCount = currentChannelStats && currentChannelStats.member_count;
+    const isTimezoneEnabled = config.ExperimentalTimezone === 'true';
```
```suggestion
    const isTimezoneEnabled = config?.ExperimentalTimezone === 'true';
```

enahum commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -408,62 +414,154 @@ export default class PostTextBoxBase extends PureComponent {
-                    actions.handleClearFiles(channelId, rootId);
+            } else if (notificationsToChannel && currentMembersCount > NOTIFY_ALL_MEMBERS && toAllorChannel) {
+                if (this.props.isTimezoneEnabled) {
+                    const {data} = await this.props.actions.getChannelTimezones(this.props.currentChannel.id);
```
definitely not keen on this one here.. it adds a delay sending the message and it will fetch it everytime, maybe this should be done during mount and when the channel changes

hahmadia replied:
Thanks for suggestion. I've made it so it only fetches on componenetdidmount and when channel changes. :)

enahum commented on app/components/combined_system_message/combined_system_message.js:
```diff
@@ -9,7 +9,6 @@ import {Posts} from 'mattermost-redux/constants';
 
 import {makeStyleSheetFromTheme} from 'app/utils/theme';
 import {t} from 'app/utils/i18n';
-
```
Please revert the change to this file

enahum commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -388,11 +411,10 @@ export default class PostTextBoxBase extends PureComponent {
 
-    sendMessage = () => {
-        const {actions, currentUserId, channelId, files, rootId} = this.props;
+    sendMessage = async () => {
```
No need for async

enahum commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -408,62 +430,146 @@ export default class PostTextBoxBase extends PureComponent {
+        const {intl} = this.context;
+        const {channelTimezoneCount} = this.state;
+        const {isTimezoneEnabled} = this.props;
+        const notifyAllTitle = (
```
This constant and the following two can be defined in the constructor as they never change

enahum commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -106,11 +113,27 @@ export default class PostTextBoxBase extends PureComponent {
             BackHandler.addEventListener('hardwareBackPress', this.handleAndroidBack);
         }
+
+        if (this.props.isTimezoneEnabled) {
```
I think you should create a function to call this action as the code here and the one in componentWillReceiveProps is practically the same.  With that you also get rid of the eslint rules problem and the need to set async to the lifecycle method.

hahmadia replied:
Thanks for the suggestion! I put all that code into a function.

enahum commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -106,11 +113,27 @@ export default class PostTextBoxBase extends PureComponent {
         if (nextProps.channelId !== this.props.channelId || nextProps.rootId !== this.props.rootId) {
-            this.setState({value: nextProps.value});
+            let channelTimezoneCount = 0;
+            if (nextProps.isTimezoneEnabled) {
```
You should also check that the previous prop is diff than the nextProp to avoid multiple calls as the nextProp can be true while other props are being updated.

Also as I mention above this can be done in a function.

On top of what I previously mentioned I’m not very keen on adding more code to this lifecycle method as it will make it harder to migrate, remember that this method is now deprecated. Try and think of a different way to achieve this.

hahmadia replied:
I did some research and came across [this](https://reactjs.org/docs/react-component.html#componentdidupdate). I have removed all the extra code I added and moved it to the `componenetDidMount` lifecycle method.

enahum commented on app/screens/channel/channel_base.js:
```diff
@@ -123,6 +126,7 @@ export default class ChannelBase extends PureComponent {
 
         if (nextProps.currentChannelId !== this.props.currentChannelId &&
             nextProps.currentTeamId === this.props.currentTeamId) {
+            nextProps.actions.getChannelStats(nextProps.currentChannelId);
```
you should move this its own if statement as this will not trigger when you switch teams.

```js
if (nextProps.currentChannelId !== this.props.currentChannelId) {
    this.props.actions.getChannelStats(nextProps.currentChannelId);
}
```

hahmadia replied:
Done! Thanks :)
